File: experiments/020-PythonVsJava/cells/python-pytest_1-Empty_free/2019-2c-parcial2-masked/20260921-213927-34997/analysis/own-tests-project/src/surveycrawler/survey_crawler.py
from surveycrawler.crawler_facing import CrawlerFacing
from surveycrawler.crawler_facing_right import CrawlerFacingRight
from surveycrawler.crawler_facing_up import CrawlerFacingUp
from surveycrawler.crawler_facing_down import CrawlerFacingDown
from surveycrawler.crawler_facing_left import CrawlerFacingLeft
from surveycrawler.firm_sand_sonar import FirmSandSonar
from surveycrawler.free_run import FreeRun


class SurveyCrawler:

    # instance creation

    @classmethod
    def at_facing(cls, a_position, a_facing_name):
        return cls.at_facing_with_sonar(a_position, a_facing_name, FirmSandSonar())

    @classmethod
    def at_facing_with_sonar(cls, a_position, a_facing_name, a_sonar):
        return cls(a_position, CrawlerFacing.facing(a_facing_name), a_sonar)

    # initialization

    def __init__(self, a_position, a_facing, a_sonar):
        self._position = a_position
        self._facing = a_facing
        self._sonar = a_sonar
        self._run = FreeRun(self)

    # exceptions

    @classmethod
    def invalid_facing_error_description(cls):
        return "Invalid facing"

    @classmethod
    def invalid_command_error_description(cls):
        return "Invalid command"

    @classmethod
    def can_not_move_over_boulder_error_description(cls):
        return "Can not move over a boulder"

    @classmethod
    def can_not_turn_over_boulder_error_description(cls):
        return "Can not turn over a boulder"

    @classmethod
    def unpredictable_movement_error_description(cls):
        return "Can not move over silt during a guarded run"

    @classmethod
    def nested_guarded_run_error_description(cls):
        return "Can not start a guarded run inside another guarded run"

    @classmethod
    def unfinished_guarded_run_error_description(cls):
        return "Guarded run was not finished"

    def signal_invalid_command(self):
        raise RuntimeError(self.invalid_command_error_description())

    def signal_can_not_move_over_boulder(self):
        raise RuntimeError(self.can_not_move_over_boulder_error_description())

    def signal_can_not_turn_over_boulder(self):
        raise RuntimeError(self.can_not_turn_over_boulder_error_description())

    def signal_unpredictable_movement(self):
        raise RuntimeError(self.unpredictable_movement_error_description())

    def signal_nested_guarded_run(self):
        raise RuntimeError(self.nested_guarded_run_error_description())

    def signal_unfinished_guarded_run(self):
        raise RuntimeError(self.unfinished_guarded_run_error_description())

    # facing

    def face_right(self):
        self._facing = CrawlerFacingRight()

    def face_up(self):
        self._facing = CrawlerFacingUp()

    def face_down(self):
        self._facing = CrawlerFacingDown()

    def face_left(self):
        self._facing = CrawlerFacingLeft()

    def turn_counter_clockwise(self):
        self.current_ground().turn_crawler_counter_clockwise(self)

    def turn_clockwise(self):
        self.current_ground().turn_crawler_clockwise(self)

    def turn_facing_counter_clockwise(self):
        self._facing.turn_counter_clockwise(self)

    def turn_facing_clockwise(self):
        self._facing.turn_clockwise(self)

    # testing

    def is_at_facing(self, a_position, a_facing_name):
        return self._position == a_position and self._facing.is_facing(a_facing_name)

    # ground

    def current_ground(self):
        return self._sonar.ground_at(self._position)

    # moving

    def advance(self):
        self.move_towards(self._facing.advance_direction())

    def retreat(self):
        self.move_towards(self._facing.retreat_direction())

    def move_towards(self, a_direction):
        a_ground = self._sonar.ground_at(self._position.plus(a_direction))
        a_ground.move_crawler_towards(self, a_direction)

    def slide_towards(self, a_direction, a_silt):
        self._run.slide_on(a_silt, a_direction)

    def displace(self, a_direction):
        self._position = self._position.plus(a_direction)

    # undoing

    def undo_advance(self):
        # the crawler retraces its own path, so the sonar is not asked again
        self.displace(self._facing.retreat_direction())

    def undo_retreat(self):
        self.displace(self._facing.advance_direction())

    def undo_turn_clockwise(self):
        self.turn_facing_counter_clockwise()

    def undo_turn_counter_clockwise(self):
        self.turn_facing_clockwise()

    # command processing

    def process(self, a_sequence_of_commands):
        FreeRun(self).process(iter(a_sequence_of_commands))

    def current_run(self):
        return self._run

    def begin_run(self, a_run):
        self._run = a_run

File: experiments/020-PythonVsJava/cells/java-gradle_1-Empty_free/2019-2c-parcial2-masked/20260921-212555-19812/analysis/acceptance-project/src/main/java/surveycrawler/SurveyCrawler.java
package surveycrawler;

public class SurveyCrawler {

    private Point position;
    private CrawlerFacing facing;
    private final Sonar sonar;
    private CrawlerRun currentRun;
    private CommandToRepeat commandToRepeat;

    // instance creation

    public static SurveyCrawler atFacing(Point aPosition, String aFacingName) {
        return atFacingSonar(aPosition, aFacingName, new FirmSandSonar());
    }

    public static SurveyCrawler atFacingSonar(Point aPosition, String aFacingName, Sonar aSonar) {
        return new SurveyCrawler(aPosition, CrawlerFacing.facing(aFacingName), aSonar);
    }

    public static String invalidFacingErrorDescription() {
        return "Invalid facing";
    }

    // exceptions

    public String invalidCommandErrorDescription() {
        return "Invalid command";
    }

    public String boulderFoundErrorDescription() {
        return "Can not move to nor turn on a boulder";
    }

    public String unpredictableMovementErrorDescription() {
        return "Can not move on silt during a guarded run";
    }

    public String guardedRunAlreadyStartedErrorDescription() {
        return "Can not start a guarded run inside a guarded run";
    }

    public String guardedRunNotStartedErrorDescription() {
        return "Can not end a guarded run that was not started";
    }

    public String guardedRunNotFinishedErrorDescription() {
        return "Guarded run was not finished";
    }

    public void signalInvalidCommand() {
        throw new RuntimeException(invalidCommandErrorDescription());
    }

    public void signalBoulderFound() {
        throw new RuntimeException(boulderFoundErrorDescription());
    }

    public void signalUnpredictableMovement() {
        throw new RuntimeException(unpredictableMovementErrorDescription());
    }

    public void signalGuardedRunAlreadyStarted() {
        throw new RuntimeException(guardedRunAlreadyStartedErrorDescription());
    }

    public void signalGuardedRunNotStarted() {
        throw new RuntimeException(guardedRunNotStartedErrorDescription());
    }

    public void signalGuardedRunNotFinished() {
        throw new RuntimeException(guardedRunNotFinishedErrorDescription());
    }

    // initialization

    private SurveyCrawler(Point aPosition, CrawlerFacing aFacing, Sonar aSonar) {
        position = aPosition;
        facing = aFacing;
        sonar = aSonar;
        currentRun = new NormalRun();
        commandToRepeat = new NoCommandToRepeat();
    }

    // facing

    public void faceRight() {
        facing = new CrawlerFacingRight();
    }

    public void faceUp() {
        facing = new CrawlerFacingUp();
    }

    public void faceDown() {
        facing = new CrawlerFacingDown();
    }

    public void faceLeft() {
        facing = new CrawlerFacingLeft();
    }

    public void turnCounterClockwise() {
        groundAtCurrentPosition().turnCrawlerCounterClockwise(this);
    }

    public void turnClockwise() {
        groundAtCurrentPosition().turnCrawlerClockwise(this);
    }

    public void turnFacingCounterClockwise() {
        rotateCounterClockwise();
        currentRun.registerMovement(new CounterClockwiseTurn());
    }

    public void turnFacingClockwise() {
        rotateClockwise();
        currentRun.registerMovement(new ClockwiseTurn());
    }

    public void rotateCounterClockwise() {
        facing.turnCounterClockwise(this);
    }

    public void rotateClockwise() {
        facing.turnClockwise(this);
    }

    // testing

    public boolean isAtFacing(Point aPosition, String aFacingName) {
        return position.equals(aPosition) && facing.isFacing(aFacingName);
    }

    public boolean isRetreatCommand(char aCommand) {
        return aCommand == 't';
    }

    public boolean isAdvanceCommand(char aCommand) {
        return aCommand == 'a';
    }

    public boolean isTurnCounterClockwiseCommand(char aCommand) {
        return aCommand == 'g';
    }

    public boolean isTurnClockwiseCommand(char aCommand) {
        return aCommand == 'h';
    }

    public boolean isRepetitionCommand(char aCommand) {
        return Character.isDigit(aCommand);
    }

    public boolean isStartOfGuardedRunCommand(char aCommand) {
        return aCommand == '(';
    }

    public boolean isEndOfGuardedRunCommand(char aCommand) {
        return aCommand == ')';
    }

    // sensing

    private Ground groundAtCurrentPosition() {
        return sonar.groundAt(position);
    }

    // moving

    public void retreat() {
        facing.retreat(this);
    }

    public void advance() {
        facing.advance(this);
    }

    public void moveRight() {
        moveTowards(new Point(1, 0));
    }

    public void moveUp() {
        moveTowards(new Point(0, 1));
    }

    public void moveDown() {
        moveTowards(new Point(0, -1));
    }

    public void moveLeft() {
        moveTowards(new Point(-1, 0));
    }

    private void moveTowards(Point aDirection) {
        sonar.groundAt(position.plus(aDirection)).moveCrawlerTowards(this, aDirection);
    }

    public void moveOneCellTowards(Point aDirection) {
        changePositionTowards(aDirection);
        currentRun.registerMovement(new Displacement(aDirection));
    }

    public void moveUnpredictablyTowards(Point aDirection, int aNumberOfCells) {
        currentRun.moveUnpredictably(this, aDirection, aNumberOfCells);
    }

    public void changePositionTowards(Point aDirection) {
        position = position.plus(aDirection);
    }

    // guarded runs

    public void beginGuardedRun() {
        currentRun = new GuardedRun();
    }

    public void finishGuardedRun() {
        currentRun = new NormalRun();
    }

    private void abortRun() {
        currentRun.undoMovementsOn(this);
        currentRun = new NormalRun();
        forgetCommandToRepeat();
    }

    // command processing

    public void process(String aSequenceOfCommands) {
        try {
            for (char aCommand : aSequenceOfCommands.toCharArray()) {
                processCommand(aCommand);
            }
            currentRun.assertIsFinished(this);
        } catch (RuntimeException anError) {
            abortRun();
            throw anError;
        }
    }

    public void processCommand(char aCommand) {
        if (isRepetitionCommand(aCommand)) {
            repeatLastCommand(aCommand);
            return;
        }
        if (isStartOfGuardedRunCommand(aCommand)) {
            currentRun.startGuardedRun(this);
            forgetCommandToRepeat();
            return;
        }
        if (isEndOfGuardedRunCommand(aCommand)) {
            currentRun.endGuardedRun(this);
            forgetCommandToRepeat();
            return;
        }

        executeCommand(aCommand);
        rememberCommandToRepeat(aCommand);
    }

    public void executeCommand(char aCommand) {
        if (isAdvanceCommand(aCommand)) {
            advance();
            return;
        }
        if (isRetreatCommand(aCommand)) {
            retreat();
            return;
        }
        if (isTurnClockwiseCommand(aCommand)) {
            turnClockwise();
            return;
        }
        if (isTurnCounterClockwiseCommand(aCommand)) {
            turnCounterClockwise();
            return;
        }

        signalInvalidCommand();
    }

    // command repetition

    public static int additionalRepetitions() {
        return 2;
    }

    public static int numberOfRepetitionsFor(char aDigit) {
        return Character.digit(aDigit, 10) + additionalRepetitions();
    }

    private void repeatLastCommand(char aDigit) {
        commandToRepeat.repeatOn(this, numberOfRepetitionsFor(aDigit));
        forgetCommandToRepeat();
    }

    private void rememberCommandToRepeat(char aCommand) {
        commandToRepeat = new LastCommandToRepeat(aCommand);
    }

    private void forgetCommandToRepeat() {
        commandToRepeat = new NoCommandToRepeat();
    }
}
